serialization.md

```diff
@@ -0,0 +1,48 @@
+# CloudEvents Serialization Profile - Version 0.1
+
+This document specifies how a CloudEvent is to be serialized into certain
+encoding formats. Compliant CloudEvents implementations that support these
+formats MUST adhere to these rules for these formats.
+
+## Table of Contents
+- [Notational Conventions](#notational-conventions)
+- [JSON](#json)
+
+## Notational Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to
+be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## JSON
+
+When serialized in [JSON](https://tools.ietf.org/html/rfc7159) a CloudEvent
+MUST adhere to the following format:
+
+```
+{
+  "cloud-events-version": "cloud-events-version value",
+  "event-id": "event-id value",
+  "namespace": "namespace value",
+  "source": {
+    "type": "source-id value",
+    "id": "source-id value"
+  },
+  "event-type": "event-type value",
+  "event-type-version": "event-type value",
+  "schema-url": "schema-url value",
+  "extensions": {
+    ... extensions values ...
+  },
+  "data": ... data value ...
```

Comment:
As an FYI: though Google&Firebase started out with a model like this where all context is at the top level with one magic field called "data", we generally considered this to be a mistake. Many languages have better ways to expose context (e.g. Golang's `Context.context`, or JavaScript's variadic parameter system) so we are migrating to a world where data & context are exposed separately/natively our runtimes. Alpha testers have given very positive feedback for this because it allows any existing function(SomeNativeType) to be a CloudFunction<SomeNativeType>.

To match this, the REST JSON is being migrated to a two-field JSON object with `{"context", "data"}`

Reply:
I could live with that. What do other people think?

Reply:
@inlined would `extensions` be under `context` or another sibling at the root level?

Reply:
@inlined where would "could-events-version" go? I can see it being useful as a top-level thing since we may choose to rename `data` or `context` in a future version, but I can also see a need for that info to be passed downstream so people know how to interpret the `context` attributes based on our spec's version number.

I do want to explore this option, it feels cleaner - would make for a great follow-on PR

Reply:
@duglin I agree with @inlined that context and data should be on the same level as he suggested. 

One more thing:
Do we need rules how to encode data, if the content-type is something binary or text/xml?

Reply:
would that be something that the definition of the content-type (e.g. json spec) would need to define?

Reply:
afaik json spec does not cover this explicitly. Most people seem to base64 encode data.

Reply:
yea, I did a quick search too and didn't find a lot. I guess we could offer some recommendations...

Reply:
Honestly I don't have the practical experience with `content-type` and `cloud-events-version` that I do with the other fields. Here's my first SWAG:
* In my opinion, `content-type` is/should be meta to the whole spec (why are we dictating the content type within a payload that's already explicitly JSON encoded?). This should be dictated in the _transport_ layer, not the encoding layer.
* `cloud-events-version` looks cleanest at the top level in JSON though if it's going to show up in the programming model it might be easier to put it in `context`.
* I have the strong opinion (weakly held) that `extensions` should be in `context`. The developer is likely to depend on these extensions and they should be easily accessible.

Reply:
@inlined My understanding is that this PR is about how to package an arbitrary cloudevent into json. I was just wondering what happens if the content-type (the context attribute) is text/xml or image/png. The result is supposed to be a valid json object. So data would need to be encoded somehow. I don't see how this can be delegated to another layer. For events that use json for their data section, there is of course no need for encoding.